File: openai-agents-ts/src/collector.ts
import { AsyncLocalStorage } from "node:async_hooks";

import type { SpanRecord } from "./types.js";

const RESERVED_PROP_KEYS = new Set([
  "kind",
  "name",
  "status",
  "duration_ms",
  "token_count",
  "finish_reason",
  "error_type",
  "triggered",
  "from_agent",
  "to_agent",
]);

type SignalHit = {
  signal: string;
  kind: string;
  anchor_kind: "span" | "interaction";
  anchor_id: string;
};

const currentCollector = new AsyncLocalStorage<RunCollector | null>();

export class RunCollector {
  workflowName: string;
  sessionId?: string;
  inputText = "";
  outputText = "";
  startedAt: string;
  endedAt?: string;
  status: "completed" | "failed" = "completed";
  spans: SpanRecord[] = [];
  runAttributes: Record<string, unknown> = {};
  userSignalHits: SignalHit[] = [];
  private openSpanStack: string[] = [];
  private pendingSpanAttrs: Record<string, Record<string, unknown>> = {};
  private startedMono = Date.now();

  constructor(opts: { workflowName: string; sessionId?: string; inputText?: string }) {
    this.workflowName = opts.workflowName;
    this.sessionId = opts.sessionId;
    this.inputText = opts.inputText ?? "";
    this.startedAt = new Date().toISOString();
  }

  static current(): RunCollector | null {
    return currentCollector.getStore() ?? null;
  }

  static runWith<T>(collector: RunCollector, fn: () => T): T {
    return currentCollector.run(collector, fn);
  }

  static enterWith(collector: RunCollector | null): void {
    currentCollector.enterWith(collector);
  }

  close(output: string, success: boolean): void {
    this.outputText = output;
    this.status = success ? "completed" : "failed";
    this.endedAt = new Date().toISOString();
  }

  get latencyMs(): number {
    return Date.now() - this.startedMono;
  }

  pushOpenSpan(spanId: string): void {
    const sid = String(spanId ?? "").trim();
    if (sid) this.openSpanStack.push(sid);
  }

  popOpenSpan(spanId: string): void {
    const sid = String(spanId ?? "").trim();
    if (!sid) return;
    if (this.openSpanStack.at(-1) === sid) {
      this.openSpanStack.pop();
      return;
    }
    this.openSpanStack = this.openSpanStack.filter((x) => x !== sid);
  }

  currentSpanId(): string | null {
    return this.openSpanStack.at(-1) ?? null;
  }

  mergeRunAttributes(attributes: Record<string, unknown>): void {
    for (const [name, value] of Object.entries(attributes)) {
      const key = String(name ?? "").trim();
      if (key) this.runAttributes[key] = value;
    }
  }

  mergePendingSpanAttributes(attributes: Record<string, unknown>): void {
    const sid = this.currentSpanId();
    if (!sid) return;
    const bucket = (this.pendingSpanAttrs[sid] ??= {});
    for (const [name, value] of Object.entries(attributes)) {
      const key = String(name ?? "").trim();
      if (key) bucket[key] = value;
    }
  }

  appendSignalHit(
    name: string,
    opts: { kind?: string; spanRef?: string | null } = {},
  ): void {
    const sig = String(name ?? "").trim();
    if (!sig) return;
    let kind = String(opts.kind ?? "context").trim();
    if (kind !== "context" && kind !== "trigger") kind = "context";
    const sid = String(opts.spanRef ?? this.currentSpanId() ?? "").trim();
    this.userSignalHits.push(
      sid
        ? { signal: sig, kind, anchor_kind: "span", anchor_id: sid }
        : { signal: sig, kind, anchor_kind: "interaction", anchor_id: "ix_0" },
    );
  }

  addSpan(span: SpanRecord): void {
    const out: SpanRecord = { ...span, properties: { ...(span.properties ?? {}) } };
    const sid = String(out.span_ref ?? "").trim();
    const pending = sid ? this.pendingSpanAttrs[sid] : undefined;
    if (pending) {
      delete this.pendingSpanAttrs[sid];
      const props = { ...(out.properties ?? {}) };
      for (const [key, value] of Object.entries(pending)) {
        if (RESERVED_PROP_KEYS.has(key) || props[key] != null) continue;
        props[key] = value;
      }
      out.properties = props;
    }
    this.spans.push(out);
  }
}

export function toInputStr(value: unknown): string {
  if (value == null) return "";
  if (typeof value === "string") return value;
  return JSON.stringify(value);
}

File: openai-agents/src/ollie_integrations_openai_agents/collector.py
"""In-memory collector for one OpenAI Agents run()."""

from __future__ import annotations

import time
from contextvars import ContextVar
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


_current_collector: ContextVar[RunCollector | None] = ContextVar("ollie_openai_run_collector", default=None)


@dataclass
class RunCollector:
    """One workflow invocation: run input/output + flat execution spans."""

    workflow_name: str
    session_id: str | None = None
    input_text: str = ""
    output_text: str = ""
    started_at: str = field(default_factory=utc_now_iso)
    ended_at: str | None = None
    status: str = "completed"
    spans: list[dict[str, Any]] = field(default_factory=list)
    _started_mono: float = field(default_factory=time.monotonic)

    @classmethod
    def current(cls) -> RunCollector | None:
        return _current_collector.get()

    @classmethod
    def set_current(cls, collector: RunCollector | None) -> None:
        _current_collector.set(collector)

    def close(self, *, output: str = "", success: bool = True) -> None:
        self.output_text = output
        self.status = "completed" if success else "failed"
        self.ended_at = utc_now_iso()

    @property
    def latency_ms(self) -> int:
        return int((time.monotonic() - self._started_mono) * 1000)

    def add_span(self, span: dict[str, Any]) -> None:
        self.spans.append(span)
